[클린코드 7기 김병호] 로또 미션 STEP 1

etff opened the pull request:
## TL;DR
로또 기능 구현

## 변경 내용
- 로또 구매 기능
- 로또 결과 출력

## 과정
- 1단계 - 콘솔 기반 로또 게임

## 코멘트

etff committed:
chore: lint 설정 추가

- eslint
- prettierrc

etff committed:
docs: 요구사항 추가

etff committed:
feat: 로또 번호 클래스 생성

etff committed:
feat: 로또 클래스 추가

etff committed:
feat: 예산 클래스 추가

etff committed:
feat: 당첨로또 클래스 추가

etff committed:
feat: 상금 클래스 추가

etff committed:
feat: 입력 클래스 추가

etff committed:
feat: 로또 번호 캐싱 추가

etff committed:
feat: 수익률 기능 추가

etff committed:
feat: 로또 게임 추가

etff committed:
feat: 로또 controller 추가

etff committed:
feat: 당첨금 누적 기능 추가

etff committed:
feat: 당첨로또 유효성 검증 추가

etff committed:
feat: 로또 게임 추가

etff committed:
feat: 로또 결과 출력 추가

etff committed:
style: 여백 제거

Tanney-102 commented on __tests__/domain/Budget.test.js:
```diff
@@ -0,0 +1,39 @@
+      const lottoPrice = 1000;
+      expect(budget.getLottoCount(lottoPrice)).toBe(10);
+    });
+  });
```
예외 케이스에 대해서도 테스트를 추가하면 어떨까요? 예를들어 new Bugdet(5001)을 호출하면 어떻게 되나요?

etff replied:
@Tanney-102 
테스트 코드에서 좀 더 명시적으로 로또 값이 리턴되는 것을 보여주면 좋을 것 같아요. 
예산 범위내에서 구매되는 로또 갯수를 리턴하는 값을 보여주도록 검증했습니다. 
``` javascript
    it("예산이 로또 가격의 배수가 아닐 때 로또 갯수는 예산을 로또 가격으로 나눈 몫이어야 한다", () => {
      const givenBudget = 5001;
      const budget = new Budget(givenBudget);
      const lottoPrice = 1000;
      expect(budget.getLottoCount(lottoPrice)).toBe(5);
    });
```

Tanney-102 commented on __tests__/domain/Budget.test.js:
```diff
@@ -0,0 +1,39 @@
+    ])(
+      "총 당첨금액이 %i원일 때, 수익률은 %s%%여야 한다",
+      (winningAmount, expectedProfit) => {
+        const givenBudget = 8000;
```
테스트 케이스 description에 givenBudget에 대한 정보도 들어가 있어야 할 듯 합니다.

초기 투입 금액이 ***원 인 경우. 총 당첨 금액이...

etff replied:
@Tanney-102 
테스트가 진행될떄 context 에 대한 맥락에 대한 정보도 있으면 좋겠네요. 해당 내용 반영했습니다.

Tanney-102 commented on src/view/InputView.js:
```diff
@@ -0,0 +1,47 @@
+        resolve(input);
+      });
+    });
+  }
```
도메인 관계 없이 공용으로 사용할 수 있는 함수는 별도의 공용 모듈로 분리해도 좋을 듯 합니다!

etff replied:
@Tanney-102 
이 부분은 util로 빼도 되겠네요. 해당 내용 반영했습니다.

Tanney-102 commented on src/view/InputView.js:
```diff
@@ -0,0 +1,47 @@
+    return this.readLineAsync("구입금액을 입력해 주세요.\n").then((input) => {
+      return new Budget(Number(input));
+    });
+  }
```
view에서 모델 인스턴스를 직접 생성하는 방식도 나쁘지는 않은데요!
view는 input을 잘 포맷팅 해서 컨트롤러에 전달만 해주고 컨트롤러에서 인스턴스 생성 등 모델과 뷰 간 커뮤니케이션을 담당하도록 역할을 잘 나눠보면 어떨까요? 이게 잘 작성만되면 나중에 view를 변경하게 될때도 조금 힘을 덜 들일 수도 있지 않을까? 하는 생각이 들기도하네요. (물론 해보기전까지는 모를 일이지만요 ㅎㅎ)

etff replied:
@Tanney-102 
사실 이 부분은 입력 부분을 만들때 항상 고민되는 부분이긴해요.
개발자에 따라 입력된 값을 바로 사용 가능한 상태의 객체라는 것을 보장받기위해,
입력과 동시에 모델 인스턴스를 직접 생성하는 방식을 선호하기도 하는 분도 계시는 것 같아요.

남겨주신 내용대로 수정해보면 입력과 객체 생성을 분리하면 뷰를 나눈 것의 장점을 더 사용해볼 수 있을 것 같아요. 
나중에 아직 어떤 상태가 될지는 잘 모르겠네요. 

예를 들어 잘못된 요구으로 인해 다시 사용자 입력을 받는 요구사항이 있다면, 이를 처리하기 위한 
scope가 조금 더 늘어날 것 같지만 아직까진 그런 요구사항이 없어서 말씀하신대로 분리해보았어요.

Tanney-102 commented on src/step1-index.js:
```diff
@@ -2,3 +2,11 @@
+import LottoController from "./controller/LottoController.js";
+
+async function main() {
+  const racingCarController = new LottoController();
+  await racingCarController.run();
```
깔끔하고 선언적인 main 함수 좋네요 👍

Tanney-102 commented on src/domain/Budget.js:
```diff
@@ -0,0 +1,20 @@
+  getProfit() {
+    return ((this.totalWinningAmount / this.amount) * 100).toFixed(1);
+  }
+}
```
이 Budget이라는 클래스가 참 애매하네요..!
일반적인 이름에 비해 사실상 하는 일은 LottoGame의 helper 같은 느낌이에요.
실제 LottoGame의 메서드에 인수로 넣어 쓰는 경우가 대부분인 것 같기도 하구요.
그럴거면 차라리 LottoGame안에서 amount, totalWinningAmount를 멤버로 가지고 있는 게 낫지 않을까요?
그럼에도 Buget의 역할을 분리하고 싶다면 LottoGame이 Buget을 멤버로 가지고 있는 형태가 적절해 보입니다.
처음 로또 구매부터 수익률 계산까지 같은 Budget 객체를 참조해야할 것 같아서요.

etff replied:
@Tanney-102 
충분히 공감되는 내용이에요. 로또 게임에 입력받은 예산, 수익, 금액 관련한 내용은 로또 게임에서 모두 처리해도 상관은 없습니다.
저의 경우는 Lotto > LottoNumber를 만들었던 것처럼, 금액적인 부분은 책임을 나눠서 처리하는게 좀더 좋아보였어요. 

이를 통해 Budget 클래스는 입력된 금액의 유효성 검증이나 예외 처리,비즈니스 규칙 적용 등을 가지고 있고, 
Budget을 분리해서 테스트하는것이 유용해보여서  클래스로 유지하는 것으로 생각해보았습니다. 
Budget 클래스를 제거하는 것도 방법이지만, 제안해주신 방향을 반영하여 LottoGame에서 참조하도록 수정해보았어요.

Tanney-102 commented on src/domain/LottoNumber.js:
```diff
@@ -0,0 +1,42 @@
+    this.#value = value;
+  }
+  equals(other) {
+    return other instanceof LottoNumber && this.#value === other.#value;
```
'#' prefix를 붙이신 건 private 멤버를 의도하신 것 같은데 다른 인스턴스(other)의 private멤버를 이렇게 참조해도 괜찮은 걸까요?

etff replied:
@Tanney-102 
`LottoNumber`를 값 객체로 사용하고 있고, 클래스 내부에서는 이와 같이 다른 인스턴스의 private 멤버를 참조하는 것이 
허용되기 때문에 가능할거라 봤어요.

참고한 소스는 다음과 같습니다.
> https://martinfowler.com/bliki/ValueObject.html
> https://github.com/eternity-oop/object/blob/master/chapter04/src/main/java/org/eternity/money/Money.java

Tanney-102 replied:
오호 js는 ts와 private 멤버에 대한 접근 범위가 다르군요..!
'가능'의 여부보다는 좋은 방식인지 고민이되네요.
말씀 주신 것처럼 값객체로 사용할 의도였다면 불변성을 위해서라도 더욱 인스턴스 단위의 캡슐화를 신경써야 하지 않을까 하는 생각도 스치구요 ㅎㅎ 이건 저도 좀 더 고민해봐야겠습니다..

etff replied:
@Tanney-102 
제가 '가능'이라고 표현한 부분은 링크의 JVM 기반언어에서
같은 클래스에서 내부 필드 비교가 가능하고 이에 대한 참고할만한 소스는 제시할 수 있기 때문인데요.

JavaScript 값 객체에 대한 equals 구현 부분에 신뢰할만한 링크는 찾았지만, 
private field를 이용한 equals 구현의 링크는 제가 찾지는 못했습니다. 

다시 돌아가서 인스턴스 멤버를 참고해도 괜찮은지는
먼저, 오류가 발생하지 않는 것은 [ECMA 스펙](https://tc39.es/ecma262/#sec-privateget)에서  정의 되었기 때문인 것으로 보입니다. 

클래스 인스턴스가 동일한 클래스의 다른 인스턴스의 개인 필드에 액세스할 수 있는 이유는 
"private" 액세스가 어떻게 정의되는지에 달라 질 것 같아요.

예를 들어 자바에서 private 접근 제어자는 개별 객체 수준이 아닌 탑레벨 클래스 수준에서 접근을 제한한다고 정의하고 있습니다.
https://docs.oracle.com/javase/specs/jls/se17/html/jls-6.html#jls-6.6.1

private 멤버의 주요 목적은 외부 코드에서 클래스의 내부 구현 세부 사항을 숨기는 캡슐화입니다.
그러나 클래스 자체에서는 서로 다른 인스턴스가 서로의 데이터와 상호 작용해야 하는 경우가 있습니다. 
예를 들어 리뷰에 남겨주신 두 객체의 내부 상태를 비교하는 equals() 가 그러한 예로 알고 있습니다.

하지만, 제가 JavaScript의 상세 스펙을 알고 있는 것은 아니라서 구글 Gemini 2.0으로 검색했을때는 
클로져를 사용한 프라이빗 변수는 인스턴스간 공유가 되지 않는다는 답변을 받았습니다.

제가 남긴 내용이 틀릴 수 있습니다. 
부족한 점은 코멘트 남겨주시면 저도 학습하는데 도움이 될 것 같습니다.

Tanney-102 commented on src/domain/LottoNumber.js:
```diff
@@ -0,0 +1,42 @@
+      throw new Error(LottoNumber.INVALID_LOTTO_NUMBER);
+    }
+    return LottoNumber.CACHE[value - LottoNumber.MIN_LOTTO_NUMBER];
+  }
```
이 메서드의 의도는 무엇인가요..?

etff replied:
@Tanney-102 
만들어두고 제대로 쓰지 않았던 것 같네요. 기존의 로또 번호를 미리 만들어두고, 만들어둔 로또 번호를 가져오기 위한 정적 메서드입니다.

Tanney-102 commented on src/domain/LottoGenerator.js:
```diff
@@ -0,0 +1,28 @@
+import LottoNumber from "./LottoNumber.js";
+import Lotto from "./Lotto.js";
+
+class LottoGenerator {
```
static 멤버만 가질거라면 class로 선언할 필요가 있나요?

etff replied:
@Tanney-102 
감사합니다. 굳이 클래스가 필요없을 것 같아서 제거했어요.

Tanney-102 commented on src/domain/Lotto.js:
```diff
@@ -0,0 +1,51 @@
+
+  getLottoNumbers() {
+    return this.#lottoNumbers;
+  }
```
배열의 레퍼런스를 그대로 반환하는 건 위험해 보입니다. 더욱이 private 멤버라면요..!

```suggestion
  getLottoNumbers() {
    return [...this.#lottoNumbers];
  }
```

etff replied:
@Tanney-102 
감사합니다. Lotto 클래스 뿐만 아니라 LottoGame에도 같은 실수가 있어서 수정해두었습니다.

Tanney-102 commented on src/domain/prize/FifthPrize.js:
```diff
@@ -0,0 +1,9 @@
+import Prize from "./Prize.js";
+
+class FifthPrize extends Prize {
+  constructor() {
+    super(3, 5000);
+  }
+}
```
constructor만 호출하는 경우라면 굳이 상속까지 해야할까 싶어요.
보너스 넘버를 확인해야하는 케이스 정도만 따로 상속해서 작성하거나 혹은 matching 전략을 따로 주입할 수 있게 해도 충분히 커버 가능해보입니다.

```ts

const defaultMatchingStrategy = (lotto, winningLotto) => {
  return winningLotto.countMatchNumbers(lotto) === this.lottoNumberMatchCount;
}

class Prize {
  constructor(..., matchingStrategy = defaultMatchingStrategy) {
    ...
    this.matchingStrategy = matchingStrategy
  }

 matched(lotto, winningLotto) {
   this.matchingStrategy(lotto, winningLotto)
 }
}
```

etff replied:
@Tanney-102 
이 부분도 충분히 제안 주실만한 내용이라고 생각했어요. 전략을 주입받는 것도 저는 좋은 방법이라고 생각해요. 

클래스 분리를 통해서 해보려고 했던 것은 다음과 같습니다.
- 좀더 구체적인 도메인 모델링
Prize, FirstPrize, 보너스 상금 등 각각의 상금 개념을 별도의 객체로 구현하면,  세분화해서 자신의 역할과 책임을 가지게 되어, 수정 및 확장이 필요할 때 관련 코드만 변경된다고 생각했어요. 현재는 단순히 constructor만 호출하는 형태일지라도, 나중에 보너스 번호 확인이나 추가 로직이 필요해지면 각 클래스에 개별적인 행동을 추가하기 쉬워질 것 같아요

단순히 map 형태로 요구사항을 구현해보는 것도 가능했지만 연습이기때문에 
객체로서 분리해보고, 이를 테스트 코드로 작성하는 것을 해보고 싶었던 것도 있습니다.

Tanney-102 reviewed (changes requested):
안녕하세요 병호님! 리뷰어 오태은입니다!
리뷰가 늦어져서 죄송합니다ㅜㅜ
step1 진행하느라 고생 많으셨어요. 도메인을 정말 촘촘하게 나누어두셨네요! 각각이 제 역할만 잘 수행하도록 고민하신 흔적이 느껴집니다. 
몇가지 개선해보면 좋을 만한 부분 코멘트 드렸으니 반영 후 다시 리뷰 요청 주시면 감사하겠습니다.

etff committed:
refactor: 리뷰 반영

- lotto game 내 budget 이동
- 전략 패턴 사용
- 배열 리턴 수정
- 로또 생성 class 제거
- view 에서 객체 생성 제거
- 읽기 함수 분리
- Budget 클래스 주석 및 예외케이스 추가

etff committed:
refactor: budget 예외 케이스 추가

etff commented:
@Tanney-102 
코드가 너무 길어서 step1부터 리뷰하기 쉽지는 않으셨을 것 같아요. 꼼꼼하게 봐주셔서 감사합니다. 
남겨주신 내용들을 반영하고, 코멘트 남겨주신 내용들에 대해서 답변을 남겨두었어요.

etff committed:
style: 숫자 표현 수정

Tanney-102 reviewed (approved):
병호님! 리뷰 반영하시느라 고생 많으셨습니다.
많은 고민 끝에 작성된 코드들을 것이라 생각은 했지만 하나하나 충분한 근거를 가지고 있는 코드임을 또 한 번 느끼게 되네요! 다음 스텝에서 뵙겠습니다!
